R/readData.R

```diff
@@ -1,19 +1,91 @@
 # ---------------------------------------------------------------------------- #
 # fast reading of big tables
-readTableFast<-function(filename,header=TRUE,skip=0,sep=""){
+readTableFast<-function(filename,header="auto",skip=0,sep="auto",as.datatable=FALSE, ...){
   
-  tab5rows <- read.table(filename, header = header,skip=skip,
-                         sep=sep, nrows = 100, stringsAsFactors=FALSE)
-  classes  <- sapply(tab5rows, class)
-  df = read.table(filename, 
-                  header = header,
-                  skip=skip,
-                  sep=sep, 
-                  colClasses = classes,
-                  stringsAsFactors=FALSE)
+  if(.Platform$OS.type=="unix"){
+    
+    #if file is in gzip, bz2 or zip format
+    if(grepl("^.*(.gz)[[:space:]]*$", filename)){
+      filename <- paste("gzip -dc", filename)
+    } else if (grepl("^.*(.zip)[[:space:]]*$", filename)){
+      filename <- paste("unzip -p", filename)
+    } else if (grepl("^.*(.bz2)[[:space:]]*$", filename)){  
+      filename <- paste("bzcat -dc", filename)
+    } 
+    
+    # removes the track header 
+    if(grepl("^.*(.gz|.bz2|.zip)[[:space:]]*$", filename)){
+      con <- pipe(filename)     
+      track=scan(con, n=1, what='character', sep='\n', quiet=TRUE)   
+      close(con)
+    }else{
+      track=scan(filename, n=1, what='character', sep='\n', quiet=TRUE) 
+    }
+    if(grepl('^>',track))
+      skip = max(1, skip)
+    
+    df = fread(filename, 
+               header=header, skip=skip,
+               sep=sep, stringsAsFactors=FALSE,
+               data.table=as.datatable)
+  }else{
+    
+    # default header and sep for read.table
+    if(header=="auto") header=FALSE
+    if(sep=="auto") sep=""
+    
+    # if file is not in zip format
+    if (!grepl("^.*(.zip)[[:space:]]*$", filename)){
+      
+      # removes the track header
+      track=scan(filename, n=1, what='character', sep='\n', quiet=TRUE)                    
+      if(grepl('^>',track))
+        skip = max(1, skip)
+      # if header = FALSE, checks whether a header line exists, and skips it
+      if(header==FALSE){
```

Comment:
I think this should be header=="auto, to be on the safe side. if the user sets header=FALSE, then we try to find a header, we will find a header if all the elements of the table is made of characters. This supposed to be a generic function, should work on all tables, not only bed-like tables